src/pdal/PyArray.hpp

```diff
@@ -77,14 +79,14 @@ class PDAL_DLL Array
     bool rowMajor() const { return m_rowMajor; };
     Shape shape() const { return m_shape; }
     const Fields& fields() const { return m_fields; };
-    ArrayIter& iterator();
+    std::shared_ptr<ArrayIter> iterator();
```

Comment:
Guess I don't understand why the iterator has been made a shared_ptr.

Reply:
I guess I thought it would make sense to transfer the ownership of the pointer to the caller, instead of maintaining a `std::vector<std::unique_ptr<ArrayIter>> m_iterators` in this class...

Had to use a shared_ptr instead of a unique_ptr though because the lambda used in setIncrementer(...) can't transfer ownership of the unique_ptr.
Maybe I should just revert it all? :)

Reply:
Do you like that better than just having `iterator()` return by value and capture by value in incrementer?

```
ArrayIter Array::iterator()
{
    return ArrayIter(m_array, m_stream_handler);
}
...
auto incrementer = [arrayIter = array->iterator(), firstPoint = true](...)
```

Reply:
Returning ArrayIter by value will make its destructor run and the numpy iterator be freed, won't it?

```
ArrayIter::~ArrayIter()
{
    if (m_iter != nullptr) {
        NpyIter_Deallocate(m_iter);
    }
}
```

Also I think it would require the copy constructor of ArrayIter to be "undeleted" for the copy to actually be possible.

As this was just a nit-picky change not quite relevant to the main purpose of the PR, I'd be more comfortable not to touch it further and just revert it back if you want.
Thanks!